test/model.test.js

```diff
@@ -8805,6 +8805,81 @@ describe('Model', function() {
       });
     });
   });
+
+  describe('castObject() (gh-11945)', function() {
+    it('casts values', function() {
+      const Test = db.model('Test', mongoose.Schema({
+        _id: false,
+        num: Number,
+        nested: {
+          num: Number
+        },
+        subdoc: {
+          type: mongoose.Schema({
+            _id: false,
+            num: Number
+          }),
+          default: () => ({})
+        },
+        docArr: [{
+          _id: false,
+          num: Number
+        }]
+      }));
+
+      const obj = {
+        num: '1',
+        nested: { num: '2' },
+        subdoc: { num: '3' },
+        docArr: [{ num: '4' }]
+      };
+      const ret = Test.castObject(obj);
+      assert.deepStrictEqual(ret, {
+        num: 1,
+        nested: { num: 2 },
+        subdoc: { num: 3 },
+        docArr: [{ num: 4 }]
+      });
+    });
+
+    it('throws if cannot cast', function() {
```

Comment:
@vkarpov15 I'm a bit hesitant whether this be the default behavior or not.

The thing is, the current alternative to castObject would be:
```js
const user = new User({ age: 'I am not a valid number' });
const castedUser = user.toObject(); // does not throw, just ignores the invalid field
```

I'm anticipating that people will start refactoring their code to use the more performant `Model.castObject(...)` and assume it works the same way as `.toObject()`.
I think it'd be nice if we can have both functions behave the same way at some point in the future.

What do you think? @vkarpov15

Reply:
But then again, there is a downside to that unified behavior:
For the `Document#toObject()`, you can always `Document#validate` to see if the document is valid or not, but for `Model.castObject()` if we don't throw an error, there won't be a simple way for people to know whether their input object is valid or not.

Not necessarily agreeing/disagreeing, thinking out loud here.

Reply:
@AbdelrahmanHafez with your example, `user.toObject()` won't throw, but `user.save()` will. It isn't that Mongoose ignores the invalid value for `age`, it just won't throw until `save()`, unless you overwrite `age` with a valid value.

It may be nice to add an option to avoid throwing, and instead just not set invalid fields.

Reply:
`user.save()` scenario is not my concern. What I'm concerned about is the rarer case where people just want to cast an object, so they use this `new Document => Document.toObject()` hack to cast the object.

An option to control error handling would be nice. 👍